[![install with bioconda](https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg?style=flat)](http://bioconda.github.io/recipes/binchicken/README.html)
[![DOI](https://zenodo.org/badge/531785064.svg)](https://zenodo.org/doi/10.5281/zenodo.10511708)
![](https://anaconda.org/bioconda/binchicken/badges/version.svg)
![](https://anaconda.org/bioconda/binchicken/badges/latest_release_relative_date.svg)
![](https://anaconda.org/bioconda/binchicken/badges/license.svg)
![](https://anaconda.org/bioconda/binchicken/badges/platforms.svg)

# Bin Chicken

[<img src="docs/_include/binchicken_logo.png" width="50%" />](binchicken_logo.png)

Bin Chicken - recovery of low abundance and taxonomically targeted metagenome assembled genomes (MAGs) through strategic coassembly.

Documentation can be found at https://AroneyS.github.io/binchicken/

Logo by Georgina H. Joyce | www.georginajoyce.com

Not to be confused with https://gitlab.com/andreassh/binchicken

## Citations

<!-- NOTE: Citations should manually be kept in sync between the repo README and the docs README -->

Aroney, S.T.N., Newell, R.J.P., Tyson, G.W. and Woodcroft B.J. _Bin Chicken: targeted metagenomic coassembly for the efficient recovery of novel genomes._ Nat Methods (2025). [https://doi.org/10.1038/s41592-025-02901-1](https://doi.org/10.1038/s41592-025-02901-1).

Bin Chicken is built on the SingleM tool, which is described in the following publication:

Woodcroft B.J., Aroney, S.T.N., Zhao, R., Cunningham, M., Mitchell, J.A.M., Nurdiansyah, R., Blackall, L. & Tyson, G.W. _Comprehensive taxonomic identification of microbial species in metagenomic data using SingleM and Sandpiper._ Nat Biotechnol (2025). [https://doi.org/10.1038/s41587-025-02738-1](https://doi.org/10.1038/s41587-025-02738-1).

If you use Aviary (through `--run-aviary`), please see the [Aviary documentation](https://github.com/rhysnewell/aviary/#Citations) for how to cite Aviary and its underlying dependencies.
